### PR TITLE
fixes #4855 - Add DECLARE statement in snowflake dialect

### LINKED ISSUE #4855: SQLFluff doesn't seem to know about a DECLARE statement on its own in Snowflake.
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

This is a mininal repro.

```
DECLARE
    results RESULTSET;
BEGIN
    results := (SELECT 1);
    RETURN TABLE(results);
END;
```

Gets:

```
> docker run -it --rm -v $PWD:/sql sqlfluff/sqlfluff:2.1.0 lint --dialect snowflake sql/repro.sql
== [sql/repro.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'DECLARE\n
                       | results RESULTSET'
L:   3 | P:   6 |  PRS | Line 3, Position 6: Found unparsable section: '\n
                       | results := (SELECT 1)'
L:   5 | P:   1 | LT02 | Line should not be indented.
                       | [layout.indent]
L:   5 | P:   5 |  PRS | Line 5, Position 5: Found unparsable section: 'RETURN
                       | TABLE(results)'
L:   6 | P:   1 |  PRS | Line 6, Position 1: Found unparsable section: 'END'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
All Finished 📜 🎉!
>   
```

### Expected Behaviour

Snowflake is happy with it, runs it fine. I expected SQLFluff to be too.

### Observed Behaviour

In What Happened.

### How to reproduce

In What Happened.

### Dialect

Snowflake.

### Version

2.1.0. Similar output in 1.4.5.

### Configuration

None. Just .sqlfluff with...

```
[sqlfluff]
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by YungChunLu**
### Brief summary of the change made

fixes #4855 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/dialects/snowflake/declare.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+BEGIN
+    profit := 1.0;
+END;
+
+DECLARE
+    res RESULTSET DEFAULT (SELECT price FROM invoices);
+    c1 CURSOR FOR res;
+BEGIN
+    RETURN c1;
+END;
```
Comment (WittierDinosaur) [inline test/fixtures/dialects/snowflake/declare.sql:15]: nit: newline

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Comment by WittierDinosaur** [comment]
linting issues

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 2]
```diff
@@ -8300,3 +8309,59 @@ class BindVariableSegment(BaseSegment):
...
+                    ),
+                    # Resultset assignment
+                    Sequence(
+                        "RESULTSET",
+                        Ref("WalrusOperatorSegment"),
+                        Bracketed(Ref("SelectableGrammar")),
+                    ),
+                    # Exception assignment
+                    Sequence(
+                        "EXCEPTION", Bracketed(Delimited(Ref("ExceptionCodeSegment"), Ref("QuotedLiteralSegment")))
```
Comment (keraion) [inline src/sqlfluff/dialects/dialect_snowflake.py:8358]: It appears `black` is still taking issue with this line.

**Review by keraion — APPROVED** [review]
LGTM.

**Pull request merged**